### PR TITLE
Use storage selector in create table path and fileio read path

### PR DESCRIPTION
**Pull request by vikrambohra**
## Summary
Primary changes
1. Use storage returned by Storage selector to select storage for a new table during creation
2. Use correct FileIO for the storage type returned from HTS for a table
3. Fixed bug where storage type always defaults to HDFS in hts

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [ ] Refactoring
- [ ] Documentation
- [x] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [x] Manually Tested on local docker setup. Please include commands ran, and their output.
- [x] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

For all the boxes checked, include a detailed description of the testing done for the changes made in this pull request.
Docker compose S3 test, Db:d3, Table:t1
1. Get table without creation -> Returns table not found with storage selected as s3 from storage selector

Screenshot 2024-09-11 at 5 16 12 PM

2. Create table -> Returns success with storage selected as s3 from storage selector, hts row with storageTyppe=s3 in hts row, minio s3 ui shows correct metadata 

Screenshot 2024-09-11 at 5 17 23 PM
Screenshot 2024-09-11 at 5 19 34 PM
Screenshot 2024-09-11 at 5 18 22 PM

3. Get table after table creation -> Returns success with storage selected as s3 from hts row

Screenshot 2024-09-11 at 5 20 37 PM

# Additional Information

- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java [hunk 1]
```diff
@@ -111,8 +114,8 @@ public TableDto save(TableDto tableDto) {
               tableIdentifier,
               writeSchema,
               partitionSpec,
-              storageManager
-                  .getDefaultStorage()
+              storageSelector
+                  .selectStorage(tableDto.getDatabaseId(), tableDto.getTableId())
```
Comment (cbb330) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java:118]: how are we able to ensure this change is safe for all implementations of `allocagteTableLocation` for example the implementation extended in li-openhouse

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java [hunk 2]
```diff
@@ -224,7 +227,8 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
           InternalCatalogMetricsConstant.METADATA_UPDATE_LATENCY);
 
       houseTable = houseTableMapper.toHouseTable(updatedMetadata);
-
+      // Set the storage type of table in house table before persisting in hts
+      houseTable.setStorageType(fileIOManager.getStorage(io()).getType().getValue());
```
Comment (jainlavina) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java:231]: I assume this io() in turn will use the storage selector for new tables when they are first committed?

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java [hunk 3]
```diff
@@ -76,14 +80,20 @@ void setup() {
...
             Mockito.mock(SnapshotInspector.class),
             mockHouseTableMapper,
             TEST_TABLE_IDENTIFIER,
-            new MetricsReporter(new SimpleMeterRegistry(), "TEST_CATALOG", Lists.newArrayList()));
+            new MetricsReporter(new SimpleMeterRegistry(), "TEST_CATALOG", Lists.newArrayList()),
+            fileIOManager);
+    LocalStorage localStorage = mock(LocalStorage.class);
+    when(fileIOManager.getStorage(fileIO)).thenReturn(localStorage);
+    when(localStorage.getType()).thenReturn(StorageType.LOCAL);
+    // when(localStorage.getType()).thenReturn(StorageType.LOCAL);
```
Comment (jainlavina) [inline iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java:96]: Remove the commented code?

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java [hunk 4]
```diff
@@ -76,14 +80,20 @@ void setup() {
     MockitoAnnotations.openMocks(this);
     Mockito.when(mockHouseTableMapper.toHouseTable(Mockito.any(TableMetadata.class)))
         .thenReturn(mockHouseTable);
+    HadoopFileIO fileIO = new HadoopFileIO(new Configuration());
```
Comment (jainlavina) [inline iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java:83]: May be we should add another test for a different FileIO given that HadoopFileIO is the default fileIO.

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java [hunk 5]
```diff
@@ -50,11 +58,12 @@ public class OpenHouseInternalCatalog extends BaseMetastoreCatalog {
   protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
     return new OpenHouseInternalTableOperations(
         houseTableRepository,
-        fileIOManager.getFileIO(storageManager.getDefaultStorage().getType()),
+        getFileIO(tableIdentifier),
         snapshotInspector,
         houseTableMapper,
         tableIdentifier,
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:64]: lets not keep FileIOManager as the last parameters, I've observed generally that metrics related params are kept last

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java [hunk 6]
```diff
@@ -120,4 +129,28 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
   }
+
+  /**
+   * Get the file IO for a table. if table exists, return the fileIO for the storageType in hts else
+   * return the fileio for storageType returned by storage selector
+   *
+   * @param tableIdentifier
+   * @return fileIO
+   */
+  private FileIO getFileIO(TableIdentifier tableIdentifier) {
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:140]: nit: getFileIO\resolveFileIO

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java (hunk 2 above)
Comment (HotSushi) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java:231]: ➕ lets use `fileIO` instance instead of `io()`

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/model/HouseTable.java [hunk 7]
```diff
@@ -10,6 +9,7 @@
 import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.NoArgsConstructor;
+import lombok.Setter;
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/model/HouseTable.java:12]: lets avoid setter pattern, we want the object to be immutable

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/model/HouseTableTest.java [hunk 8]
```diff
@@ -33,9 +32,7 @@ public void testHouseTableDefaultValues() {
             continue;
           }
 
-          if (fieldName.equals("storageType")) {
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/model/HouseTableTest.java:36]: The right way would be to make the mapper take fileIO as an input and during the mapping, use that fileIO to get the right type

File: services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java [hunk 9]
```diff
@@ -111,8 +114,8 @@ public TableDto save(TableDto tableDto) {
               tableIdentifier,
               writeSchema,
               partitionSpec,
-              storageManager
-                  .getDefaultStorage()
+              storageSelector
```
Comment (HotSushi) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java:117]: do we make a log in each storage selection call? if not lets ensure we do!

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java (hunk 4 above)
Comment (HotSushi) [reply iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java:83]: we could add add a test for HadoopFileIO with empty prefix and localFileIO with prefix: `file:`

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java [hunk 10]
```diff
@@ -224,7 +227,8 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
           InternalCatalogMetricsConstant.METADATA_UPDATE_LATENCY);
 
       houseTable = houseTableMapper.toHouseTable(updatedMetadata);
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java:229]: could we change the method here to ` houseTable = houseTableMapper.toHouseTable(updatedMetadata, fileIO)`  to keep object immutability and use mapping logic to correctly set FileIO

**Review by HotSushi — CHANGES_REQUESTED** [review]

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java (hunk 6 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:140]: done

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java (hunk 5 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:64]: done

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java (hunk 2 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java:231]: done

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java [hunk 11]
```diff
@@ -40,6 +44,10 @@ public class OpenHouseInternalCatalog extends BaseMetastoreCatalog {
 
   @Autowired StorageManager storageManager;
 
+  @Autowired StorageSelector storageSelector;
```
[the remark this answers is not in the document]
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:54]: StorageSelector is used to select storage for a table at the time of table creation. It has a method selectStorage(db,tbl) that can be used to  select a storage other than the cluster default (storageManager.getDefaultStorage).

File: services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java (hunk 1 above)
Comment (vikrambohra) [reply services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java:118]: Currently the only selector implementation we have is the DefaultStorageSelector that defaults to the cluster level storage so this change should have no impact.

There are multiple unit tests that test the behavior of allocateTableLocation which would fail if the behavior has changed

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/model/HouseTable.java (hunk 7 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/model/HouseTable.java]: Lets discuss in https://github.com/linkedin/openhouse/pull/187#discussion_r1750715536

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java (hunk 10 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java]: Lets discuss in https://github.com/linkedin/openhouse/pull/187#discussion_r1750715536

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/model/HouseTableTest.java (hunk 8 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/model/HouseTableTest.java:36]: Updated. PTAL.

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java [hunk 12]
```diff
@@ -50,6 +51,8 @@ public class OpenHouseInternalTableOperations extends BaseMetastoreTableOperatio
 
   HouseTableRepository houseTableRepository;
 
+  FileIOManager fileIOManager;
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java:54]: fileIOManager needed anymore?

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java [hunk 13]
```diff
@@ -138,4 +148,36 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
+    try {
+      houseTable =
+          houseTableRepository.findById(
+              HouseTablePrimaryKey.builder()
+                  .databaseId(tableIdentifier.namespace().toString())
+                  .tableId(tableIdentifier.name())
+                  .build());
+    } catch (HouseTableNotFoundException e) {
+      log.info(
+          "House table not found {}.{}",
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java:170]: nit: house table entry not found

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/mapper/HouseTableMapper.java [hunk 14]
```diff
@@ -4,22 +4,30 @@
...
 public abstract class HouseTableMapper {
+  @Autowired FileIOManager fileIOManager;
+
   @Mapping(target = "lastModifiedTime", ignore = true)
   @Mapping(target = "creationTime", ignore = true)
-  public abstract HouseTable toHouseTable(Map<String, String> properties);
+  @Mapping(
+      target = "storageType",
+      expression = "java(fileIOManager.getStorage(fileIO).getType().getValue())")
+  public abstract HouseTable toHouseTable(Map<String, String> properties, FileIO fileIO);
```
Comment (HotSushi) [inline iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/mapper/HouseTableMapper.java:27]: nice! is there a unit test for this?

**Review by HotSushi — COMMENTED** [review]
Can we run a docker test with s3 and inside that test, ensure HTS row contains "s3" as an entry?

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java (hunk 12 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperations.java]: yes not needed. removed

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/mapper/HouseTableMapper.java (hunk 14 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/mapper/HouseTableMapper.java:27]: done

File: iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java (hunk 13 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/main/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalCatalog.java]: done

**Comment by vikrambohra** [comment]
> Can we run a docker test with s3 and inside that test, ensure HTS row contains "s3" as an entry?

S3 test added in testing section

File: iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java (hunk 4 above)
Comment (vikrambohra) [reply iceberg/openhouse/internalcatalog/src/test/java/com/linkedin/openhouse/internal/catalog/OpenHouseInternalTableOperationsTest.java:83]: currently added s3 docker tests in testing section. We ca address this in a future PR

**Review by HotSushi — APPROVED** [review]
LGTM
